### PR TITLE
[AuthZ] Add IRequirementData

### LINKED ISSUE #44551: AuthZ: Add IAuthorizationRequirementData
## Background and Motivation

We want to make it easier to attribute endpoints to add additional requirements for authorization.  This interface will help by letting metadata or custom attributes specify requirements which will automatically get combined with other existing authorization mechanisms.  Note the presence of IAuthorizationRequirementData metadata will be treated similar to IAuthorizeData, meaning its considered opting in for authorization (but it does not automatically bring in the default policy, see examples)

## Proposed API

```diff
namespace Microsoft.AspNetCore.Authorization;

+ public interface IAuthorizationRequirementData
{
+    IEnumerable<IAuthorizationRequirement> GetRequirements();
}
```

## Usage Examples

Example 1: Derive from Authorize attribute without specifying a policy to customize the default policy by adding arbitrary requirements, i.e.
```csharp
internal class MinimumAgeAuthorizeAttribute : AuthorizeAttribute, IAuthorizationRequirementData
{
    public MinimumAgeAuthorizeAttribute(int age) => Age = age;

    public int Age { get; }

    IEnumerable<IAuthorizationRequirement> GetRequirements() => new[] { 
        new MinimumAgeRequirement(Age)
    };
}
    
    [MinimumAgeAuthorize(21)]
    public void NoKidsControllerMethod { ... }
```

Example 2: Fully specify the requirements without interacting with default policy (no authorize attribute), i.e. only allow authorization on Tuesdays

```csharp
internal class OnlyTuesdayAttribute : Attribute, IAuthorizationRequirementData
{
    IEnumerable<IAuthorizationRequirement> GetRequirements() => new[] { 
        new OnlyTuesdayRequirement()
    };
}
    
    [OnlyTuesday]
    public void TacoTuesdayMethod { ... }
```

## Alternative Designs

Decided against allowing specification of a fully policy to minimize the complexity of how it combines with existing methods of producing policies, as its trivial to combine additional requirements (always safe to combine)

## Risks

New way to opt into authorization where before only [Authorize]/`IAuthorizeData` was used to require authorization.

### PR DESCRIPTION
**Pull request by HaoK**
Explore allowing an attribute to add requirements

### CODE REVIEW

**Comment by HaoK** [comment]
@davidfowl 

I went back to just enabling adding requirements via an attribute since that makes it a little simpler to combine as there already are existing ways to specify what authentication, i.e. `[Authorize]`.

**Comment by davidfowl** [comment]
Maybe we just need `IRequirementData` to exist. We don't need `RequiresAttribute`. I'm thinking the typical use case for this is what we have [documented here today](https://learn.microsoft.com/en-us/aspnet/core/security/authorization/iauthorizationpolicyprovider?view=aspnetcore-6.0#custom-authorization-attributes). After this change, it would look like this:

```C#
internal class MinimumAgeAuthorizeAttribute : AuthorizeAttribute, IRequirementData
{
    public MinimumAgeAuthorizeAttribute(int age) => Age = age;

    public int Age { get; }

    IEnumerable<IAuthorizationRequirement> GetRequirements() => new[] { 
        new AuthorizationPolicyBuilder().AddRequirements(new MinimumAgeRequirement(Age)).Build()  
    };
}
```

**Comment by davidfowl** [comment]
> Sounds good, but in your example, GetRequirements would be even simpler right, since we don't need to build the policy anymore, it'd just be: new[] { new MinimumAgeRequirement(Age) } without the policy builder


Oops, fixed.

**Comment by HaoK** [comment]
Okay updated, pretty tiny change that actually enables a lot

**Comment by HaoK** [comment]
Note: IRequirementData does trigger authorization by itself, so it doesn't require an `[Authorize]` attribute or IAuthorizeData to have an effect

File: src/Security/Authorization/test/AuthorizationMiddlewareTests.cs [hunk 1]
```diff
@@ -374,6 +376,55 @@ public async Task CanApplyPolicyDirectlyToEndpoint()
         Assert.True(calledPolicy);
     }
 
+    public class ReqAttribute : Attribute, IRequirementData
```
Comment (davidfowl) [inline src/Security/Authorization/test/AuthorizationMiddlewareTests.cs:379]: Add an example that does:

```C#
public class ReqAttribute2 : AuthorizeAttribute, IRequirementData
```

Comment (HaoK) [reply src/Security/Authorization/test/AuthorizationMiddlewareTests.cs:379]: Updated and added the CanApplyAuthorizeRequirementAttributeDirectlyToEndpoint test below for this

**Review by Tratcher — APPROVED** [review]

**Comment by davidfowl** [comment]
Lets do API review, file an issue and lets get this checked in and marked for blogging 😄 cc @adityamandaleeka

**Comment by kevinchalet** [comment]
May I suggest naming it `IAuthorizationRequirementData` in case you'd come with a similar design for other areas like CORS in the future?

**Comment by HaoK** [comment]
Sounds reasonable I'll update the design issue with the suggested new name, but I won't update this PR until after the api review has been done in case of other feedback

**Pull request merged**